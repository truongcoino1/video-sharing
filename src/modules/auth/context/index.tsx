"use client";

import { ReactNode } from 'react'
import { AuthProvider} from './AuthContext'

const AllProviders = ({ children }: { children: ReactNode }) => {
  return <AuthProvider>{children}</AuthProvider>
}

export {AllProviders}
export * from './AuthContext'