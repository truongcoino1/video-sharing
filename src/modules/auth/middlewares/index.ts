import { Middleware } from '../types/middleware'

export const withMiddlewares = (...middlewares: Middleware[]): Middleware => {
  // Create chain of middlewares
  const chain = middlewares.reduceRight(
    (next, middleware) => (req, res) => middleware(req, res, next),
    (_req, res) => res.status(200).json({ success: true })
  )
  return chain
}
