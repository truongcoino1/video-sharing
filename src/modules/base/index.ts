export * from './types/api.d'
export * from './services'
export * from './hooks'