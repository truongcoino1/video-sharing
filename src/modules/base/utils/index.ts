export function flushPromises(timeout = 0): Promise<unknown> {
  return new Promise(function (resolve) {
    setTimeout(resolve, timeout);
  });
}

export const uniqueArrayById = (array: any[]) => {
  return array.filter((v, i, a) => a.findIndex((t) => t.id === v.id) === i && !!v.id);
};
