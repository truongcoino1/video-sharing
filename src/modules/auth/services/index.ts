import { api, ApiResponse } from "@/modules/base";
import { User } from "../types/auth";

export const AuthService = {
  login: async (
    email: string,
    password: string
  ): Promise<ApiResponse<User>> => {
    return await api.post("/api/auth/login", { email, password });
  },
};
