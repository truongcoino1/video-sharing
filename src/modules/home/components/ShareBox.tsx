import React from "react";
import { useInput, Button } from "@/modules/base";
import { useMovies } from "../hooks/useMovies";
import { useAuthContext } from "@/modules/auth";
import SkeletonShareBox from "./SkeletonShareBox";

const ShareBox = () => {
  const { value: url, handleOnChange: handleSetUrl } = useInput("");
  const { shareMovie, isSharing } = useMovies();
  const { loading } = useAuthContext();

  if (loading) {
    return <SkeletonShareBox/>
  }

  return (
    <div className="sharebox-container mt-[100px] flex items-center justify-center">
      <form
        className="sharebox-form border relative border-gray-300 max-w-[calc(100vw_-_32px)] rounded-large w-[600px] p-24"
        onSubmit={(e) => {
          e.preventDefault();          
          shareMovie(url || "");
        }}
      >
        <span className="sharebox-title absolute top-[-20px] bg-light-100 font-medium text-24 px-8">
          Share a Youtube movie
        </span>
        <div className="sharebox-url flex items-center font-medium my-24">
          <label className="mr-12" htmlFor="url">
            Youtube URL:
          </label>
          <input
            autoFocus
            name="url"
            type="text"
            className="border border-gray-300 rounded-large flex-1 p-8"
            value={url}
            onChange={handleSetUrl}
          />
        </div>
        <div className="sharebox-submit flex items-center justify-center">
          <Button isLoading={isSharing} className="w-full p-8" label="Share" name="submit" type="submit" isDisable={!url} />
        </div>
      </form>
    </div>
  );
};

export default ShareBox;
