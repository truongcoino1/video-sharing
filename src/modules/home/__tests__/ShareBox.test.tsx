import { render, cleanup } from "@testing-library/react";
import ShareBox from "../components/ShareBox";
import {flushPromises} from '@/modules/base';
import { AuthContext } from "@/modules/auth";
//@ts-ignore
global.setImmediate = jest.useRealTimers;

const setUpLogin = () => {
  return render(
    <AuthContext.Provider
      value={{
        currentUser: {
          email: "truongvk",
          token: "123",
          id: 1,
          password: "123",
          refreshToken: "123",
        },
        setLoading: jest.fn(),
        setCurrentUser: jest.fn(),
        loading: false,
      }}
    >
      <ShareBox />
    </AuthContext.Provider>
  );
};

const setUpLoginLoading = () => {
  return render(
    <AuthContext.Provider
      value={{
        currentUser: {
          email: "truongvk",
          token: "123",
          id: 1,
          password: "123",
          refreshToken: "123",
        },
        setLoading: jest.fn(),
        setCurrentUser: jest.fn(),
        loading: true,
      }}
    >
      <ShareBox />
    </AuthContext.Provider>
  );
};

const shareMovie = jest.fn();
jest.mock("../hooks/useMovies", () => {
  const useMovies = () => {
    return {
      shareMovie: shareMovie,
    };
  };
  return {
    useMovies,
  };
});

describe("ShareBox", () => {
  afterEach(cleanup);
  it("ShareBox should render without crash", () => {
    const result = setUpLogin();
    const elm = result.container.querySelector(".sharebox-container");
    expect(elm).toBeInTheDocument();
  });

  it("ShareBox should render form", () => {
    const result = setUpLogin();
    const elm = result.container.querySelector(".sharebox-form");
    expect(elm).toBeInTheDocument();
  });

  it("ShareBox should render title", () => {
    const result = setUpLogin();
    const elm = result.container.querySelector(".sharebox-title");
    expect(elm).toBeInTheDocument();
    expect(elm?.innerHTML).toBe("Share a Youtube movie");
  });

  it("ShareBox should render url input", () => {
    const result = setUpLogin();
    const elm = result.container.querySelector(".sharebox-url");
    expect(elm).toBeInTheDocument();
    expect(elm?.querySelector("label")?.innerHTML).toBe("Youtube URL:");
    expect(elm?.querySelector("input")?.getAttribute("name")).toBe("url");
    expect(elm?.querySelector("input")?.getAttribute("type")).toBe("text");
  });

  it("ShareBox should render submit button", () => {
    const result = setUpLogin();
    const elm = result.container.querySelector(".sharebox-submit");
    expect(elm).toBeInTheDocument();
    expect(elm?.querySelector("button")?.getAttribute("name")).toBe("submit");
    expect(elm?.querySelector("button")?.getAttribute("type")).toBe("submit");
    expect(elm?.querySelector("button")?.innerHTML).toBe("Share");
  });

  it("ShareBox should disable submit button when url is empty", () => {
    const result = setUpLogin();
    const elm = result.container.querySelector(".sharebox-submit");
    expect(elm?.querySelector("button")?.getAttribute("disabled")).toBe("");
  });

  it("ShareBox should enable submit button when url is not empty", () => {
    const result = setUpLogin();
    const elm = result.container.querySelector(".sharebox-submit");
    const input = result.container.querySelector(".sharebox-url input");
    input?.setAttribute("value", "https://www.youtube.com/watch?v=123");
    expect(elm?.querySelector("button")?.getAttribute("disabled")).toBe("");
  });

  it("ShareBox should call shareMovie when submit form", async () => {
    const result = setUpLogin();
    const form = result.container.querySelector(".sharebox-form");
    const input = result.container.querySelector(".sharebox-url input");
    input?.setAttribute("value", "https://www.youtube.com/watch?v=123");
    form?.dispatchEvent(new Event("submit", { bubbles: true }));
    await flushPromises();
   expect(shareMovie).toBeCalledTimes(1);
   
  });

  it("should render correct ShareBox", () => {
    const element = setUpLogin();
    expect(element).toMatchSnapshot();
  });

  it("should render correct ShareBox when loading", () => {
    const result = setUpLoginLoading();
    const sharebox = result.container.querySelector(".sharebox-container-skeleton");
    expect(sharebox).toBeInTheDocument();
  });
});
