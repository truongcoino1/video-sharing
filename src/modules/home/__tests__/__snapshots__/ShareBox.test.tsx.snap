// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShareBox should render correct ShareBox 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sharebox-container mt-[100px] flex items-center justify-center"
      >
        <form
          class="sharebox-form border relative border-gray-300 max-w-[calc(100vw_-_32px)] rounded-large w-[600px] p-24"
        >
          <span
            class="sharebox-title absolute top-[-20px] bg-light-100 font-medium text-24 px-8"
          >
            Share a Youtube movie
          </span>
          <div
            class="sharebox-url flex items-center font-medium my-24"
          >
            <label
              class="mr-12"
              for="url"
            >
              Youtube URL:
            </label>
            <input
              class="border border-gray-300 rounded-large flex-1 p-8"
              name="url"
              type="text"
              value=""
            />
          </div>
          <div
            class="sharebox-submit flex items-center justify-center"
          >
            <button
              class="w-full p-8 flex items-center rounded-medium justify-center opacity-50 cursor-not-allowed bg-dark-200 text-dark-high "
              disabled=""
              name="submit"
              type="submit"
            >
              Share
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sharebox-container mt-[100px] flex items-center justify-center"
    >
      <form
        class="sharebox-form border relative border-gray-300 max-w-[calc(100vw_-_32px)] rounded-large w-[600px] p-24"
      >
        <span
          class="sharebox-title absolute top-[-20px] bg-light-100 font-medium text-24 px-8"
        >
          Share a Youtube movie
        </span>
        <div
          class="sharebox-url flex items-center font-medium my-24"
        >
          <label
            class="mr-12"
            for="url"
          >
            Youtube URL:
          </label>
          <input
            class="border border-gray-300 rounded-large flex-1 p-8"
            name="url"
            type="text"
            value=""
          />
        </div>
        <div
          class="sharebox-submit flex items-center justify-center"
        >
          <button
            class="w-full p-8 flex items-center rounded-medium justify-center opacity-50 cursor-not-allowed bg-dark-200 text-dark-high "
            disabled=""
            name="submit"
            type="submit"
          >
            Share
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
